robot/web/request_counter.go

```diff
@@ -139,24 +174,187 @@ func (rc *RequestCounter) Stats() any {
+			ifarr *inFlightAndRejectedRequests,
+		) bool {
+			if ifr := ifarr.inFlightRequests.Load(); ifr > 0 {
+				ci.InFlightRequests[resourceName] = ifr
+			}
+			if rr := ifarr.rejectedRequests.Load(); rr > 0 {
+				ci.RejectedRequests[resourceName] = rr
+			}
+
+			return true
+		})
+	}
+
+	// If client has no in-flight requests, and connected >= 5 minutes ago, prune the client
+	// from the `rc.requestsForPC` and `rc.pcToClientMetadata` maps so the maps do not grow
+	// unboundedly. That client will re-appear in debug information if it makes another
+	// request.
+	if len(ci.InFlightRequests) == 0 && timeSinceConnect > 5*time.Minute {
+		rc.requestsPerPC.Delete(pc)
+		rc.pcToClientMetadata.Delete(pc)
+	}
+
+	return ci
+}
+
+// Logs that a particular request limit was exceeded. Outputs the following information
+// (where possible):
+// - Which API method invocation was attempted
+// - Which resource the API method was invoked upon
+// - Which "offending" client caused the limit to be reached (name and IP)
+// - The in-flight and rejected request counts of all connected clients
+func (rc *RequestCounter) logRequestLimitExceeded(
+	apiMethodString, resource string,
+	pc *webrtc.PeerConnection,
+) {
+	offendingClientInformation := fmt.Sprintf(
+		"%+v",
+		rc.createClientInformationFromPC(pc),
+	)
+	allOtherClientInformation := "["
+	rc.requestsPerPC.Range(func(
+		pcKey *webrtc.PeerConnection,
+		_ *ssync.Map[string, *inFlightAndRejectedRequests],
+	) bool {
+		// Do not include offending client.
+		if pc != pcKey {
+			allOtherClientInformation += fmt.Sprintf(
+				"%+v",
+				rc.createClientInformationFromPC(pcKey),
+			)
+		}
+		return true
+	})
+	allOtherClientInformation += "]"
+
+	msg := fmt.Sprintf("Request limit exceeded for resource. See %s for troubleshooting steps", ReqLimitExceededURL)
+	rc.logger.Warnw(
+		msg,
+		"method", apiMethodString,
+		"resource", resource,
+		"offending_client_information", offendingClientInformation,
+		"all_other_client_information", allOtherClientInformation,
+	)
+}
+
 // UnaryInterceptor returns an incoming server interceptor that will pull method information and
 // optionally resource information to bump the request counters.
 func (rc *RequestCounter) UnaryInterceptor(
 	ctx context.Context, req any, info *googlegrpc.UnaryServerInfo, handler googlegrpc.UnaryHandler,
 ) (resp any, err error) {
 	apiMethod := extractViamAPI(info.FullMethod)
+	pc, pcSet := rpc.ContextPeerConnection(ctx)
+	if pcSet {
+		rc.setClientMetadataForPC(ctx, pc)
+	}
+
 	if resource := buildResourceLimitKey(req, apiMethod); resource != "" {
-		if ok := rc.incrInFlight(resource); !ok {
-			rc.logger.Warnw(fmt.Sprintf("Request limit exceeded for resource. See %s for troubleshooting steps", ReqLimitExceededURL),
-				"method", apiMethod.full,
-				"resource", resource)
+		if ok := rc.incrInFlight(resource, pc); !ok {
+			rc.logRequestLimitExceeded(apiMethod.full, resource, pc)
 			return nil, &RequestLimitExceededError{
```

Comment:
I meant as part of the error here, whether it'd be useful to return the number of inflight requests from this client, so callers have an idea whether they are causing the issue. though ok to table as a future idea for now